docs/lab/lab04/lab04.md

```diff
@@ -0,0 +1,384 @@
+
+### Counters
+
+A counter is a piece of hardware logic that counts, as its name suggests. Every clock cycle, it increments the value of a register, until it overflows and starts anew. 
+
+:::info
+A regular counter on 8 bits would count up from 0 to 255, then loop back to 0 and continue counting. 
+:::
+
+TODO: insert image from GTKWave
+
+On the RP2040, the counter is associated with 3 registers:
+1. `value` - the current value of the counter
+2. `direction` - whether the counter is counting UP or DOWN
+3. `reset` - if the direction is UP, the value at which the counter resets to 0; if the direction is DOWN, the value at which the counter reset after reaching 0
+
+![Counter](images/counter.svg)
+
+The way the counter works here is that it increments/decrements every clock cycle and checks whether or not it has reached its reset value. If is has, then it resets to its initial value and starts all over again.
+
+The ARM Cortex-M uses the SysTick time counter to keep track of time. This counter is decremented every microsecond, and when it reaches 0, it triggers an exception and then resets.
+
+`SYST_CVR` register = the value of the timer itself
+
+`SYST_RVR` register = the reset value
+
+`SYST_CSR_SET` register:
+	- `ENABLE` field = enable/disable the counter
+	- `TICKINT` field = enable/disable exception on reaching 0
+
+### Timers
+
+Until now, we have been able to blink a led at a certain time interval, by waiting a while between each led toggle. The technique we used so far was asking the processor to skip a clock cycle a number of times, or by calling the processor instruction "nop" (no operation) in a loop. 
+:::info
+This method is not ideal, since the "nop" instruction stalls the processor and wastes valuable time that could otherwise be used to do other things in the meantime. To optimize this, we can use *alarms*.
+:::
+An **alarm** is a counter that triggers an interrupt every time it reaches a certain value. This way, an alarm can be set to trigger after a specific interval of time, and while it's waiting, the main program can continue executing instructions, and so it is not blocked. When the alarm reaches the chosen value, it goes off and triggers an interrupt that can then be handled in its specific ISR.
+
+![Alarm](images/alarm.svg)
+
+:::info
+The RP2040 timer is fully monotomic, meaning it can never truly overflow. Its value is stored on 64 bits and increments every 1 microsecond, which means that the last value it can increment to before overflowing is 2^64-1, which the equivalent of roughly 500.000 years. This timer allows 4 different alarms, which can be used independently (TIMER_IRQ_0/1/2/3).
```

Comment:
```suggestion
The RP2040 timer is fully monotomic, meaning it can never truly overflow. Its value is stored on 64 bits and increments every 1 microsecond, which means that the last value it can increment to before overflowing is 2<sup>64-1</sup>, which the equivalent of roughly 500,000 years. This timer allows 4 different alarms, which can be used independently (`TIMER_IRQ_0/1/2/3`).
```